lib/semantics/mod-file.cc

```diff
@@ -590,84 +608,116 @@ std::ostream &PutLower(std::ostream &os, const std::string &str) {
+    close(fd);
+    unlink(path.c_str());
+  }
+  int fd;
+  std::string path;
+};
+
+// Create a temp file in the same directory and with the same suffix as path.
+// Return an open file descriptor and its path.
+static Temp MkTemp(const std::string &path) {
+  auto length{path.length()};
+  auto dot{path.find_last_of("./")};
+  std::string suffix{dot < length && path[dot] == '.' ? path.substr(dot) : ""};
+  CHECK(length > suffix.length() &&
+      path.substr(length - suffix.length()) == suffix);
+  auto tempPath{path.substr(0, length - suffix.length()) + "XXXXXX" + suffix};
+  int fd{mkstemps(&tempPath[0], suffix.length())};
+  auto mask{umask(0777)};
+  umask(mask);
+  chmod(tempPath.c_str(), 0666 & ~mask);  // temp is created with mode 0600
+  return Temp{fd, tempPath};
+}
+
+// Write the module file at path, prepending header. If an error occurs,
+// return an error message, otherwise nullptr.
+static int WriteFile(const std::string &path, const std::string &contents) {
+  auto header{std::string{ModHeader::bom} + ModHeader::magic +
+      CheckSum(contents) + ModHeader::terminator};
+  if (FileContentsMatch(path, header, contents)) {
+    return 0;
+  }
+  Temp temp{MkTemp(path)};
+  if (temp.fd < 0) {
+    return errno;
+  }
+  if (write(temp.fd, header.c_str(), header.size()) == -1 ||
+      write(temp.fd, contents.c_str(), contents.size()) == -1) {
+    return errno;
   }
-  stream << header << '\n' << contents;
-  stream.close();
-  return !stream.fail();
+  if (std::rename(temp.path.c_str(), path.c_str()) == -1) {
+    return errno;
+  }
+  return 0;
 }
 
 // Return true if the stream matches what we would write for the mod file.
-static bool FileContentsMatch(std::fstream &stream, const std::string &header,
-    const std::string &contents) {
-  char c;
-  for (std::size_t i{0}; i < header.size(); ++i) {
-    if (!stream.get(c) || c != header[i]) {
-      return false;
-    }
+static bool FileContentsMatch(const std::string &path,
+    const std::string &header, const std::string &contents) {
+  std::size_t hsize{header.size()};
+  std::size_t csize{contents.size()};
+  if (GetFileSize(path) != hsize + csize) {
+    return false;
   }
-  if (!stream.get(c) || c != '\n') {
+  int fd{open(path.c_str(), O_RDONLY)};
+  if (fd < 0) {
     return false;
   }
-  for (std::size_t i{0}; i < contents.size(); ++i) {
-    if (!stream.get(c) || c != contents[i]) {
+  constexpr std::size_t bufSize{4096};
+  std::string buffer(bufSize, '\0');
+  if (read(fd, &buffer[0], hsize) == -1 ||
+      std::memcmp(&buffer[0], &header[0], hsize) != 0) {
+    close(fd);
+    return false;  // header doesn't match
+  }
+  for (auto remaining{csize};;) {
+    auto bytes{std::min(bufSize, remaining)};
+    auto got{read(fd, &buffer[0], bytes)};
+    if (got != static_cast<ssize_t>(bytes) ||
+        std::memcmp(&buffer[0], &contents[csize - remaining], bytes) != 0) {
+      close(fd);
       return false;
     }
+    if (got == 0 && remaining == 0) {
```

Comment:
Or just return right after `remaining` reaches 0

```
remaining -= bytes;
if (remaining == 0) {
      close(fd);
      return true;
}
```
It's not too different, just avoids one call to `read()`  and this if check gets simpler.

Reply:
I think that will be wrong if `remaining` is 0 but there are still bytes in the file to be read.

Reply:
In that case, in the next iteration
`auto bytes = std::min(4096, 0)`
`auto got = read(fd, &buffer[0], 0)` 
Nothing with be read and read() will return 0.

Reply:
Probably needs an EOF() check to make sure `contents` is not just matching a prefix of the file.

Reply:
Right, but I don't know a better way to check for EOF than to do a read and get back 0. That's what will happen in the next iteration if we keep going when `remaining` is 0 but `got` is not.